### DISCUSSION TITLE
Null or Empty Operators

### DISCUSSION THREAD

**Discussion by dhughes55** [dotnet/csharplang#5450] [Language Ideas]
### PROPOSAL
C# currently supports the following null-related operators:
- null conditional operator `?.`
- null coalescing operator `??`
- null coalescing assignment operator `??=`

These are incredibly useful operators.  However, they're not as useful when you want to check a string to see if it's null or empty. I propose that two new operators be added:
- null or empty coalescing operator `??_`
- null or empty coalescing assignment operator `??_=`

### VALUE

This new syntax will significantly shorten the amount of typing needed when you need to check a string for a null or empty value.  It's simple and intuitive.

**BEFORE:**
```
model.Comment1 = string.IsNullOrEmpty(comment1) ? "N/A" : comment1;
if (string.IsNullOrEmpty(model.Comment2)) model.Comment2 = "N/A"
```

**AFTER:**
```
model.Comment1 = comment1 ??_ "N/A";
model.Comment2 ??_= "N/A";
```

### DRAWBACKS

The only drawback I can think of is that the `??_` and `??_=` operators would look similar to their existing counterparts `??` and `??=`.  But that's also what makes it so intuitive in my opinion.

> **UPDATE:** As noted in the comment section below:
> ??_ is already legal syntax meaning ?? _.  This means the ??_ operator [no space] could lead to confusion with ?? _ [with a space].
> https://github.com/dotnet/csharplang/discussions/5450#discussioncomment-1672908

### ALTERNATIVES:

A function called `ValueOrNull()` function could also be introduced that returns null if a value is null or empty:
```
model.Comment3 = comment3.ValueOrNull() ?? "N/A";
```
Another alternative is that instead of "is null or empty" it could be used for "is null or whitespace".  However, it seems to me that this would be unnecessary since the new syntax could support whitespace easily enough using `Trim():`
```
model.Comment4 = comment4?.Trim() ??_ "N/A";
```

**Comment by HaloFour** [comment]
Why does this need an operator?  Do we need another one for "null or whitespace"?

**Reply by dhughes55** [reply]
The purpose of the operator would be to extend the existing functionality of ?? and ??= which is useful but isn't as useful for strings where a value may be null or empty.  As mentioned in the "alternatives" section of the proposal, there would be little need for a "null or whitespace" operator as the new syntax plus Trim() would handle that nicely:  `model.Comment4 = comment4?.Trim() ??_ "N/A";`

**Reply by HaloFour** [reply]
Seems you could do the same for the "null or empty" case without the need of a specialized operator.

**Comment by DavidArno** [comment]
You can already do:
```c#
model.Comment1 = comment1 is null or "" ? "N/A" : comment1;
model.Comment2 = comment2 switch {
    null or "" => "N/A",
    var c => c
};
if (model.Comment2 is null or "")) ...
```
Given the power of patterns these days, it's hard to see how yet another hard-to-remember operator, that only says a few characters, can be justified.

**Reply by dhughes55** [reply]
Thanks for the input.  In terms of length, while I agree that
```
model.Comment1 = comment1 ??_ "N/A";
```
isn't much shorter than
```
model.Comment1 = comment1 is null or "" ? "N/A" : comment1;
```
I'd respectfully submit that it's considerably tidier than using a `switch` or an `if` statement.  Which isn't to take anything away from the new pattern-based switch syntax which is quite nifty.  But pattern matching is a mixed bag: some people like it, others don't.  For those who already use `??` and `??=` regularly, `??_` and `??_=` feels like a perfectly natural syntax addition, at least to me.

**Comment by svick** [comment]
> A function called `ValueOrNull()` function could also be introduced that returns null if a value is null or empty:
>
> ```
> model.Comment3 = comment3.ValueOrNull() ?? "N/A";
> ```

To me, this sounds like a much better option than adding a confusing string-only operator. Though I would rename it to `NullIfEmpty()` (and then you could easily also add `NullIfEmptyOrWhitespace()`).

**Reply by dhughes55** [reply]
Thanks for your input!

**Comment by theunrepentantgeek** [comment]
Creating new C# syntax to enshrine treating **null** and `""` as identical seems like fertile source of bugs to me. 

In many cases (though admittedly not all), treating **null** and `""` as identical leads to subtle bugs and data corruption. 

This is common enough that many (at least in my experience) treat stray **null** string values as an error condition.

**Reply by dhughes55** [reply]
Hi-ya. While I certainly see your point, whether or not null and string.Empty _should_ be treated as identical is a separate topic.  The language _already does_ offer the ability to treat them as identical by saying `string.IsNullOrEmpty(str)`.  My proposal simply takes functionality that already exists in C# and makes it more terse, and terser code typically leads to fewer bugs, not more.

**Reply by theunrepentantgeek** [reply]
> terser code typically leads to fewer bugs, not more.

Increasing the _clarity_ of code leads to fewer bugs. Sometimes this happens when the code is terser, but the reverse happens when the code becomes so compressed it becomes cryptic. 
￼
To split a hair, `string.IsNullOrEmpty(str)` is a library feature, not a language one. 😉 

With the introduction of nullable types, the code I've seen falls into three groups. 

- **string** where **null** is not expected. Treating **null** as `""` just conceals the problem. 

- **string?** where **null** has a different meaning to `""`. Treating **null** as `""` is simply wrong. 

- **string?** where **null** has the same meaning as `""`. 

In my experience the last of those is vanishingly rare, and not common enough to require dedicated syntax.

**Reply by dhughes55** [reply]
> Increasing the clarity of code leads to fewer bugs. Sometimes this happens when the code is terser, but the reverse happens when the code becomes so compressed it becomes cryptic.

I agree wholeheartedly.  That's why it's typically a code smell to have a bunch of one-letter variables and function names, nested ternary statements, and so on.

> To split a hair, string.IsNullOrEmpty(str) is a library feature, not a language one. 😉

Touche. 😉

Don't you love nullable types?  (It's a shame the language didn't start out with them, but I'm definitely a fan all the same!)  That said, a lot of folks aren't using nullable types so there's still a lot of code out there where nulls and empty strings are used interchangeably.  Some devs always initialize their strings with an empty string to lessen the risk of null reference exceptions so in their universe, they're interchangeable.  To your point, though, it's certainly debatable whether adding a new feature that treats null and empty interchangeably is a good or bad thing.

**Comment by jnm2** [comment]
`??_` is already legal syntax meaning `?? _`.

`a??_(b)` could then mean either `a ?? (_(b))` or `a ??_ (b)`.

**Reply by dhughes55** [reply]
An excellent point.  The last two lines below look very similar:
```
string _message = "Here's a message.";
string _defaultMessage = "Here's the default.";
return _message ?? _defaultMessage;
return _message ??_ defaultMessage;
```
A point I've added to the drawbacks section.  Thanks!

**Reply by jnm2** [reply]
It's not just a drawback. It's a total blocker because it would change the meaning of code that compiles today.

**Reply by jnm2** [reply]
Unless maybe the identifier name `_` was to have been made a hard compile error for at least one major version first... and even then...

**Comment by thegyorf** [comment]
As someone going through updating a project to enforce nullable reference types I would definitely appreciate a more terse way to do "null if empty/whitespace" checks for strings.

Personally my preference would be to add an even stricter option to the nullable project setting that would tell the compiler to treat empty or whitespace as null when ?? is used against a string.
